src/logic/map_objects/pinned_note.cc

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (C) 2022 by the Widelands Development Team
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "logic/map_objects/pinned_note.h"
+
+#include <memory>
+
+#include "graphic/rendertarget.h"
+#include "logic/game_data_error.h"
+#include "logic/player.h"
+#include "wui/interactive_player.h"
+
+namespace Widelands {
+
+class PinnedNoteDescr : public BobDescr {
+public:
+	PinnedNoteDescr(char const* const init_name, char const* const init_descname)
+	   : BobDescr(init_name,
+	              init_descname,
+	              MapObjectType::PINNED_NOTE,
+	              MapObjectDescr::OwnerType::kTribe) {
+	}
+
+	Bob& create_object() const override {
+		return *new PinnedNote();
+	}
+
+private:
+	DISALLOW_COPY_AND_ASSIGN(PinnedNoteDescr);
+};
+
+static const PinnedNoteDescr g_descr("pinned_note", "Pinned Note");
+
+PinnedNote::PinnedNote() : Bob(g_descr) {
+}
+
+// static
+PinnedNote& PinnedNote::create(EditorGameBase& egbase,
+                               Player& owner,
+                               Coords pos,
+                               const std::string& text,
+                               const RGBColor& rgb) {
+	PinnedNote& note = *new PinnedNote();
+	note.set_text(text);
+	note.set_rgb(rgb);
+	note.set_owner(&owner);
+	note.set_position(egbase, pos);
+	note.init(egbase);
+	return note;
+}
+
+void PinnedNote::init_auto_task(Game& game) {
+	if (get_position().field->get_immovable() != nullptr &&
+	    get_position().field->get_immovable()->descr().type() >= MapObjectType::BUILDING) {
+		/* If the player builds a building here, move one field out of the way. */
+		FCoords new_pos = game.map().br_n(get_position());
```

Comment:
I understand the new position would be the flag position - Correct? Wouldn't it be better to move to the left?

Reply:
If it's a big building then the left would also be occupied by it so the note couldn't be there either. I chose to move the note to the flag because that's the position that is most clearly related to a building, but it's arbitrary which spot to choose so long as no building is occupying it.

Reply:
ok tested it and it looks ok